q2_quality_filter/_filter.py

```diff
@@ -332,101 +526,51 @@ def q_score(
-        ]
-    )
-
-    for barcode_id, sample_id in enumerate(demux_manifest_df.index.values):
-        # get/create filepath(s) of input/output fastq file(s)
-        forward_input_fp = demux_manifest_df.loc[sample_id, 'forward']
-        forward_output_fp = Path(result.path) / Path(forward_input_fp).name
-        if paired:
-            reverse_input_fp = demux_manifest_df.loc[sample_id, 'reverse']
-            reverse_output_fp = Path(result.path) / Path(reverse_input_fp).name
-
-        # open output filehandle(s) and create fastq record iterator
-        forward_fh = gzip.open(forward_output_fp, mode='wb')
-
-        if paired:
-            reverse_fh = gzip.open(reverse_output_fp, mode='wb')
-
-            forward_iterator = _read_fastq_records(str(forward_input_fp))
-            reverse_iterator = _read_fastq_records(str(reverse_input_fp))
-            iterator = zip(forward_iterator, reverse_iterator)
-        else:
-            iterator = _read_fastq_records(str(forward_input_fp))
-
-        for fastq_record in iterator:
-            if paired:
-                forward_record, reverse_record = fastq_record
-            else:
-                forward_record = fastq_record
-                reverse_record = None
-
-            # process records
-            forward_record, forward_status = _process_record(
-                fastq_record=forward_record,
-                phred_offset=phred_offset,
-                min_quality=min_quality,
-                window_length=quality_window + 1,
-                min_length_fraction=min_length_fraction,
-                max_ambiguous=max_ambiguous
-            )
-            reverse_record, reverse_status = _process_record(
-                fastq_record=reverse_record,
-                phred_offset=phred_offset,
-                min_quality=min_quality,
-                window_length=quality_window + 1,
-                min_length_fraction=min_length_fraction,
-                max_ambiguous=max_ambiguous
-            )
-
-            # see if record(s) retained and update filtering stats
-            retained = _is_retained(
-                forward_status,
-                reverse_status,
-                filtering_stats_df,
-                sample_id
+    # create per-sample arguments for parallel invocations
+    sample_ids = demux_manifest_df.index
+    parameters = {
+        'paired': paired,
+        'output_format_path': Path(result.path),
+        'demux_manifest': demux_manifest_df,
+        'phred_offset': phred_offset,
+        'min_quality': min_quality,
+        'quality_window': quality_window,
+        'min_length_fraction': min_length_fraction,
+        'max_ambiguous': max_ambiguous,
+    }
+    per_sample_arguments = [
+        [sample_id] + list(parameters.values()) for sample_id in sample_ids
+    ]
+
+    # schedule samples to processes
+    with multiprocessing.Pool(num_processes) as pool:
+        all_sample_stats = pool.starmap(_process_sample, per_sample_arguments)
```

Comment:
Why are you building this dictionary here? It seems like the only thing that's it's being used for is casting its values to a list for generation of the per-sample arguments. Since those are never changing within a run, it might be clearer to use `functools.partial`. For example, something like this (which assumes an import of `functools` earlier in this file):

```suggestion
    _process_sample_partial = functools.partial(_process_sample,
        paired=paired,
        output_format_path=Path(result.path),
        demux_manifest=demux_manifest_df,
        phred_offset=phred_offset,
        min_quality=min_quality,
        quality_window=quality_window,
        min_length_fraction=min_length_fraction,
        max_ambiguous=max_ambiguous)
        
    # schedule samples to processes
    with multiprocessing.Pool(num_processes) as pool:
        all_sample_stats = pool.starmap(_process_sample_partial, sample_ids)
```